include/nuttx/lib/stdatomic.h

```diff
@@ -62,6 +62,109 @@
+                         __nx_atomic_flag_test_and_set_8(obj, type))
+
+#define atomic_flag_test_and_set(obj) atomic_flag_test_and_set_n(obj, __ATOMIC_RELAXED)
+#define atomic_flag_test_and_set_explicit(obj, type) atomic_flag_test_and_set_n(obj, 1, type)
+#define atomic_flag_clear(obj) atomic_store(obj, 0)
+#define atomic_flag_clear_explicit(obj, type) atomic_store_explicit(obj, 0, type)
+
+#define atomic_fetch_and_n(obj, val, type) \
+  (sizeof(*(obj)) == 1 ? __nx_atomic_fetch_and_1(obj, val, type) : \
+   sizeof(*(obj)) == 2 ? __nx_atomic_fetch_and_2(obj, val, type) : \
+   sizeof(*(obj)) == 4 ? __nx_atomic_fetch_and_4(obj, val, type) : \
+                         __nx_atomic_fetch_and_8(obj, val, type))
+
+#define atomic_fetch_and(obj, val) atomic_fetch_and_n(obj, val, __ATOMIC_RELAXED)
+#define atomic_fetch_and_explicit(obj, val, type) atomic_fetch_and_n(obj, val, type)
+
+#define atomic_fetch_or_n(obj, val, type) \
+  (sizeof(*(obj)) == 1 ? __nx_atomic_fetch_or_1(obj, val, type) : \
+   sizeof(*(obj)) == 2 ? __nx_atomic_fetch_or_2(obj, val, type) : \
+   sizeof(*(obj)) == 4 ? __nx_atomic_fetch_or_4(obj, val, type) : \
+                         __nx_atomic_fetch_or_8(obj, val, type))
+
+#define atomic_fetch_or(obj, val) atomic_fetch_or_n(obj, val, __ATOMIC_RELAXED)
+#define atomic_fetch_or_explicit(obj, val, type) atomic_fetch_or_n(obj, val, type)
+
+#define atomic_fetch_xor_n(obj, val, type) \
+  (sizeof(*(obj)) == 1 ? __nx_atomic_fetch_xor_1(obj, val, type) : \
+   sizeof(*(obj)) == 2 ? __nx_atomic_fetch_xor_2(obj, val, type) : \
+   sizeof(*(obj)) == 4 ? __nx_atomic_fetch_xor_4(obj, val, type) : \
+                         __nx_atomic_fetch_xor_8(obj, val, type))
+
+#define atomic_fetch_xor(obj, val) atomic_fetch_xor_n(obj, val, __ATOMIC_RELAXED)
+#define atomic_fetch_xor_explicit(obj, val, type) atomic_fetch_xor_n(obj, val, type)
+
+#define atomic_fetch_add_n(obj, val, type) \
+  (sizeof(*(obj)) == 1 ? __nx_atomic_fetch_add_1(obj, val, type) : \
+   sizeof(*(obj)) == 2 ? __nx_atomic_fetch_add_2(obj, val, type) : \
+   sizeof(*(obj)) == 4 ? __nx_atomic_fetch_add_4(obj, val, type) : \
+                         __nx_atomic_fetch_add_8(obj, val, type))
+
+#define atomic_fetch_add(obj, val) atomic_fetch_add_n(obj, val, __ATOMIC_RELAXED)
+#define atomic_fetch_add_explicit(obj, val, type) atomic_fetch_add_n(obj, val, type)
+
+#define atomic_fetch_sub_n(obj, val, type) \
+  (sizeof(*(obj)) == 1 ? __nx_atomic_fetch_sub_1(obj, val, type) : \
+   sizeof(*(obj)) == 2 ? __nx_atomic_fetch_sub_2(obj, val, type) : \
+   sizeof(*(obj)) == 4 ? __nx_atomic_fetch_sub_4(obj, val, type) : \
+                         __nx_atomic_fetch_sub_8(obj, val, type))
+
+#define atomic_fetch_sub(obj, val) atomic_fetch_sub_n(obj, val, __ATOMIC_RELAXED)
+#define atomic_fetch_sub_explicit(obj, val, type) atomic_fetch_sub_n(obj, val, type)
+
+#else
+
 #define atomic_store_n(obj, val, type) \
   (sizeof(*(obj)) == 1 ? __atomic_store_1(obj, val, type) : \
```

Comment:
let's always map to to nx_ version regardless CONFIG_LIBC_ARCH_NXATOMIC

Reply:
@xiaoxiang781216 

Thank you for your advice.

I tried the mappling always nx_ version at raspberrypi-pico:nsh, the result is a lot of link error such as : mq_open.c:322: undefined reference to `__atomic_fetch_add_2'.

I think,
to impliment the mappling always nx_ version, all processor which does not support atomic instruction naturally such as ARCH_CORTEXM0 have to use 'select LIBC_ARCH_NXATOMIC'.
gcc built-in atomic functions are not supported for such processor and there is a lot of such processor,
it it not effective to apply `select LIBC_ARCH_NXATOMIC' for all such processor.
so I limited the scope of nx_ version to only CXD32XX.

Reply:
let's always map atomic_xxx to nx_atomic_xxx in header file, but implement both nx_atomic_xxx and __atomic_xxx.